fl4health/model_bases/ensemble_base.py

```diff
@@ -0,0 +1,111 @@
+        for both voting and averaging prediction of individual models.
+
+        Args:
+            ensemble_models (Dict[str, nn.Module]): A dictionary of models that make up the ensemble.
+            aggregation_mode (Optional[EnsembleAggregationMode]): The mode in which to aggregate the
+                predictions of indivdual models.
+
+        Raises:
+            ValueError: Keys of the ensemble_dicts must be valid python identifiers.
+
+        """
+        super().__init__()
+
+        self.ensemble_models = nn.ModuleDict(ensemble_models)
+        self.aggregation_mode = aggregation_mode
+
+    def forward(self, input: torch.Tensor) -> Dict[str, torch.Tensor]:
+        """
+        Produce the predictions of the ensemble models given input data.
+
+        Args:
+            input (torch.Tensor): A batch of input data.
+
+        Returns:
+            Dict[str, torch.Tensor]: A dictionary of predictions of the individual ensemble models
+                as well as prediction of the ensemble as a whole.
+        """
+        preds = {}
+        for key, model in self.ensemble_models.items():
+            preds[key] = model(input)
+
+        # Don't store gradients of when computing ensemble predictions
+        with torch.no_grad():
+            if self.aggregation_mode == EnsembleAggregationMode.AVERAGE:
+                ensemble_pred = self.ensemble_average(list(preds.values()))
+            else:
+                ensemble_pred = self.ensemble_vote(list(preds.values()))
+
+        preds["ensemble-pred"] = ensemble_pred
+
+        return preds
+
+    def ensemble_vote(self, preds_list: List[torch.Tensor]) -> torch.Tensor:
+        """
+        Produces the aggregated prediction of the ensemble via voting. Expects predictions
+        to be in a format where the 0 axis represents the sample index and the -1 axis represents
+        the class dimension.
+
+        Args:
+            preds_list (List[torch.Tensor]): A list of predictions of the models in the ensemble.
+
+        Returns:
+            torch.Tensor: The vote prediction of the ensemble.
+        """
+        assert all(preds.shape == preds_list[0].shape for preds in preds_list)
+        preds_dimension = list(preds_list[0].shape)
+
+        # If larger than two dimensions, we map to 2D to perfom voting operation (and reshape later)
+        if len(preds_dimension) > 2:
+            preds_list = [preds.reshape(-1, preds_dimension[-1]) for preds in preds_list]
+
+        # For each model prediction, compute the argmax of the model over the classes and stack column-wise into matrix
+        # Each row of matrix represents the argmax of each model for a given sample
+        argmax_per_model = torch.hstack([torch.argmax(preds, dim=1, keepdim=True) for preds in preds_list])
+        # For each row (sample), compute the unique class predictions and their respective counts
+        index_count_list = map(lambda x: torch.unique(x, return_counts=True), argmax_per_model.unbind())
```

Comment:
Wasn't sure how to do this entirely in torch because unique returns a tuple. Feel free to suggest something more efficient.

Reply:
I think this looks pretty good